src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.HillClimbing.cs

```diff
@@ -0,0 +1,385 @@
+            private readonly double _threadMagnitudeMultiplier;
+            private readonly int _sampleIntervalHigh;
+            private readonly double _throughputErrorSmoothingFactor;
+            private readonly double _gainExponent;
+            private readonly double _maxSampleError;
+
+            private double _currentControlSetting;
+            private int _totalSamples;
+            private int _lastThreadCount;
+            private double _averageThroughputNoise;
+            private double _secondsElapsedSinceLastChange;
+            private double _completionsSinceLastChange;
+            private int _accumulatedCompletionCount;
+            private double _accumulatedSampleDuration;
+            private double[] _samples;
+            private double[] _threadCounts;
+            private int _currentSampleInterval;
+            
+            private Random _randomIntervalGenerator = new Random();
+
+            public HillClimbing(int wavePeriod, int maxWaveMagnitude, double waveMagnitudeMultiplier, int waveHistorySize, double targetThroughputRatio,
+                double targetSignalToNoiseRatio, double maxChangePerSecond, double maxChangePerSample, int sampleIntervalLow, int sampleIntervalHigh,
+                double errorSmoothingFactor, double gainExponent, double maxSampleError)
+            {
+                _wavePeriod = wavePeriod;
+                _maxThreadWaveMagnitude = maxWaveMagnitude;
+                _threadMagnitudeMultiplier = waveMagnitudeMultiplier;
+                _samplesToMeasure = wavePeriod * waveHistorySize;
+                _targetThroughputRatio = targetThroughputRatio;
+                _targetSignalToNoiseRatio = targetSignalToNoiseRatio;
+                _maxChangePerSecond = maxChangePerSecond;
+                _maxChangePerSample = maxChangePerSample;
+                _sampleIntervalLow = sampleIntervalLow;
+                _sampleIntervalHigh = sampleIntervalHigh;
+                _throughputErrorSmoothingFactor = errorSmoothingFactor;
+                _gainExponent = gainExponent;
+                _maxSampleError = maxSampleError;
+
+                _samples = new double[_samplesToMeasure];
+                _threadCounts = new double[_samplesToMeasure];
+
+                _currentSampleInterval = _randomIntervalGenerator.Next(_sampleIntervalLow, _sampleIntervalHigh);
+            }
+
+            public (int newThreadCount, int newSampleInterval) Update(int currentThreadCount, double sampleDuration, int numCompletions)
+            {
+
+                //
+                // If someone changed the thread count without telling us, update our records accordingly.
+                // 
+                if (currentThreadCount != _lastThreadCount)
+                    ForceChange(currentThreadCount, State.Initializing);
+
+                //
+                // Update the cumulative stats for this thread count
+                //
+                _secondsElapsedSinceLastChange += sampleDuration;
```

Comment:
Similar to what you did for `_secondsElapsedSinceLastChange` it would be nice to indicate the units for `sampleDuration` and other such variables in its name